### PR TITLE
refactor: Exa.Trkx configurable stages

### PR DESCRIPTION
**Pull request by benjaminhuth**
This refactors the Exa.TrkX Plugin in a way that the functionality is broken up into several stages: Graph construction, Edge classification, Track building.

This should both be more flexible and also in the current state should make the code more concise due to code reusage etc.

Finally, this should allow to configure the Algorithm as follows:
```python
  s.addAlgorithm(
      acts.examples.TrackFindingAlgorithmExaTrkX(
          level=customLogLevel(),
          inputSpacePoints="spacepoints",
          outputProtoTracks="protoTracks",
          graphConstructor=acts.examples.TorchMetricLearning(**metricLearningCfg),
          edgeClassifiers=[
              acts.examples.TorchEdgeClassifier(**filterCfg),
              acts.examples.TorchEdgeClassifier(**gnnCfg),
          ],
          trackBuilder=acts.examples.BoostTrackBuilding(),
      )
  )
```

Three base classes are defined:
```c++
class GraphConstructionBase {
 public:
  // returns nodes, edges
  virtual std::tuple<std::any, std::any> operator()(std::vector<float> &inputValues) = 0;
};

class EdgeClassificationBase {
 public:
  // returns nodes, edges, edge weights
  virtual std::tuple<std::any, std::any, std::any> operator()(std::any nodes, std::any edges) = 0;
};

class TrackBuildingBase {
 public:
  // returns tracks
  virtual std::vector<std::vector<int>> operator()(
      std::any nodes, std::any edges, std::any edgeWeights, std::vector<int> &spacepointIDs) = 0;
};
```

I decided to use `std::any` as a representation for intermediate data, since I'm not sure how else to make a common type for `at::Tensor` (torch) and `Ort::Value` (Onnx) with allowing them to be compile-able separately from each other.

However I think this is not performance critical at the moment, since the chain is only executed once per event.

Should go in after #1894 to ensure the behaviour does not change.

### CODE REVIEW

**Comment by xju2** [comment]
Hi @benjaminhuth ,

This is a very nice re-factorize plan! 

Having these algorithms factorized will inevitably introduce overheads due to functional calls and data movements/conversions. But we could minimize the data movements/conversions. For examples, the `GraphConstructionBase` can simply return edges. The node indices are conventionally defined as the same as the spacepoint indices. Following that convention will save us some space. So `GraphConstructionBase` becomes
```cpp
class GraphConstructionBase {
 public:
  // returns edges
  virtual std::vector<int64> operator()(std::vector<float> &inputValues) const = 0;
};
```
The returned vector can be reshaped to (2, number of edges). 

`EdgeClassificationBase` will not modify the nodes and edges. It can be 
```cpp 
class EdgeClassificationBase {
 public:
  // returns  edge weights
  virtual std::vector<float> operator()(std::any nodes, std::any edges) const = 0;
};
```

In this way, the output data objects for these algorithms are standardized. We can also make the input data object more specific. For example, `std::any nodes`  can be changed to `std::vector<float> nodes`. In the end, it becomes the `TrackFindingAlgorithmExaTrkX`'s job to convert them into either `Ort::Value` or `at::Tensor`.

**Comment by benjaminhuth** [comment]
Hi @xju2,
I have not worked on this for some time now due to other more urgent problems, but I will try to get it in now soon. It is a very good suggestion to only return `edges` (graph construction) or `scores` (edge classification), which I will try to implement!

Regarding the `std::vector` based interface, I'm not sure yet. My intent was to minimize copying data between CPU and GPU by keeping e.g. the `torch::Tensor` in the `std::any` (so hopefully under the hood the `score`-tensor and `edge`-tensor never leave the GPU and always reference the same device-data).
Of course this will break interchange-ability between e.g. ONNX and torch based chains, but I'm not sure how desirable this is.

Let me know what you think about this!

File: Plugins/ExaTrkX/src/BoostTrackBuilding.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  // timeInfo.labeling = timer.stopAndGetElapsedTime();
+  // timeInfo.total = totalTimer.stopAndGetElapsedTime();
+
+  // if (recordTiming) {
+  //   return timeInfo;
+  // } else {
+  //   return std::nullopt;
+  // }
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/BoostTrackBuilding.cpp:121]: dead code?

File: Plugins/ExaTrkX/src/BoostTrackBuilding.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  // print_current_cuda_meminfo(logger);
+
+  if (trackLabels.size() == 0) {
+    // if (recordTiming) {
+    //   return timeInfo;
+    // } else {
+    //   return std::nullopt;
+    // }
+    return {};
+  }
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/BoostTrackBuilding.cpp:88]: dead code?

File: Plugins/ExaTrkX/src/OnnxEdgeClassifier.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+namespace Acts {
+
+OnnxEdgeClassifier::OnnxEdgeClassifier(const Config &cfg) : m_cfg(cfg) {
+  m_env = std::make_unique<Ort::Env>(ORT_LOGGING_LEVEL_WARNING,
+                                     "ExaTrkX - edge classifier");
+
+  Ort::SessionOptions session_options;
+  session_options.SetIntraOpNumThreads(1);
+  // OrtStatus* status =
+  // OrtSessionOptionsAppendExecutionProvider_CUDA(session_options, 0);
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/OnnxEdgeClassifier.cpp:27]: dead code?

File: Plugins/ExaTrkX/src/OnnxEdgeClassifier.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+  //   std::fstream out(filtering_outname, out.out);
+  //   if (!out.is_open()) {
+  //     ACTS_ERROR("failed to open " << filtering_outname);
+  //   } else {
+  //     std::copy(fOutputCTen.data_ptr<float>(),
+  //               fOutputCTen.data_ptr<float>() + fOutputCTen.numel(),
+  //               std::ostream_iterator<float>(out, " "));
+  //   }
+  // }
+  // std::cout << fOutputCTen.slice(0, 0, 3) << std::endl;
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/OnnxEdgeClassifier.cpp:100]: dead code?

File: Plugins/ExaTrkX/src/OnnxMetricLearning.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+namespace Acts {
+
+OnnxMetricLearning::OnnxMetricLearning(const Config& cfg) : m_cfg(cfg) {
+  m_env = std::make_unique<Ort::Env>(ORT_LOGGING_LEVEL_WARNING,
+                                     "ExaTrkX - metric learning");
+
+  Ort::SessionOptions session_options;
+  session_options.SetIntraOpNumThreads(1);
+  // OrtStatus* status =
+  // OrtSessionOptionsAppendExecutionProvider_CUDA(session_options, 0);
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/OnnxMetricLearning.cpp:27]: dead code?

File: Plugins/ExaTrkX/src/OnnxMetricLearning.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    std::vector<float>& inputValues, const Logger& logger) {
+  Ort::AllocatorWithDefaultOptions allocator;
+  auto memoryInfo = Ort::MemoryInfo::CreateCpu(
+      OrtAllocatorType::OrtArenaAllocator, OrtMemType::OrtMemTypeDefault);
+
+  // printout the r,phi,z of the first spacepoint
+  // ACTS_INFO("First spacepoint information: " << inputValues.size() << "\n\t");
+  // for (size_t i = 0; i < 3; i++) {
+  //   ACTS_INFO("\t" << inputValues[i]);
+  // }
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/OnnxMetricLearning.cpp:70]: dead code?

File: Plugins/ExaTrkX/src/OnnxMetricLearning.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  }
+  // if (debug) {
+  //   std::fstream out(embedding_outname, out.out);
+  //   if (!out.is_open()) {
+  //     ACTS_ERROR("failed to open " << embedding_outname);
+  //   } else {
+  //     std::copy(eOutputData.begin(), eOutputData.end(),
+  //               std::ostream_iterator<float>(out, " "));
+  //   }
+  // }
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/OnnxMetricLearning.cpp:107]: dead code?

File: Plugins/ExaTrkX/src/OnnxMetricLearning.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+  // if (debug) {
+  //   std::fstream out(edgelist_outname, out.out);
+  //   if (!out.is_open()) {
+  //     ACTS_ERROR("failed to open " << edgelist_outname);
+  //   } else {
+  //     std::copy(edgeList.begin(), edgeList.end(),
+  //               std::ostream_iterator<int64_t>(out, " "));
+  //   }
+  // }
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/OnnxMetricLearning.cpp:132]: dead code?

File: Plugins/ExaTrkX/src/TorchEdgeClassifier.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+  torch::Tensor filterMask = fOutput > m_cfg.cut;
+  torch::Tensor edgesAfterF = edgeList.index({Slice(), filterMask});
+  edgesAfterF = edgesAfterF.to(torch::kInt64);
+  const int64_t numEdgesAfterF = edgesAfterF.size(1);
+
+  ACTS_VERBOSE("Size after filter cut: " << numEdgesAfterF)
+  // print_current_cuda_meminfo(logger);
+
+  // timeInfo.filtering = timer.stopAndGetElapsedTime();
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/TorchEdgeClassifier.cpp:72]: dead code?

File: Plugins/ExaTrkX/src/TorchEdgeClassifier.cpp [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    results.back().sigmoid_();
+  }
+
+  auto fOutput = torch::cat(results);
+  results.clear();
+
+  ACTS_VERBOSE("Size after filtering network: " << fOutput.size(0));
+  ACTS_VERBOSE("Slice of filtered output:\n"
+               << fOutput.slice(/*dim=*/0, /*start=*/0, /*end=*/9));
+  // print_current_cuda_meminfo(logger);
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/TorchEdgeClassifier.cpp:62]: dead code?

File: Plugins/ExaTrkX/src/TorchEdgeClassifier.cpp [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+std::tuple<std::any, std::any, std::any> TorchEdgeClassifier::operator()(
+    std::any inputNodes, std::any inputEdges, const Logger& logger) {
+  const torch::Device device(m_deviceType);
+
+  const auto eLibInputTensor = std::any_cast<torch::Tensor>(inputNodes);
+  const auto edgeList = std::any_cast<torch::Tensor>(inputEdges);
+
+  c10::InferenceMode guard(true);
+  // timer.start();
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/TorchEdgeClassifier.cpp:41]: dead code?

File: Plugins/ExaTrkX/src/TorchMetricLearning.cpp [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  ACTS_VERBOSE("First spacepoint information [r, phi, z]: "
+               << inputValues[0] << ", " << inputValues[1] << ", "
+               << inputValues[2]);
+  ACTS_VERBOSE("Max and min spacepoint: "
+               << *std::max_element(inputValues.begin(), inputValues.end())
+               << ", "
+               << *std::min_element(inputValues.begin(), inputValues.end()))
+  // print_current_cuda_meminfo(logger);
+
+  // ExaTrkXTimer timer(not recordTiming);
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/TorchMetricLearning.cpp:52]: dead code?

File: Plugins/ExaTrkX/src/TorchMetricLearning.cpp [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+  eInputTensorJit.push_back(eLibInputTensor.to(device));
+  std::optional<at::Tensor> eOutput =
+      e_model.forward(eInputTensorJit).toTensor();
+  eInputTensorJit.clear();
+
+  ACTS_VERBOSE("Embedding space of the first SP:\n"
+               << eOutput->slice(/*dim=*/0, /*start=*/0, /*end=*/1));
+  // print_current_cuda_meminfo(logger);
+
+  // timeInfo.embedding = timer.stopAndGetElapsedTime();
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/TorchMetricLearning.cpp:76]: dead code?

File: Plugins/ExaTrkX/src/TorchMetricLearning.cpp [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+  // ****************
+  // Building Edges
+  // ****************
+
+  // timer.start();
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/TorchMetricLearning.cpp:82]: dead code?

File: Plugins/ExaTrkX/src/TorchMetricLearning.cpp [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+  std::optional<torch::Tensor> edgeList = buildEdges(
+      *eOutput, numSpacepoints, m_cfg.embeddingDim, m_cfg.rVal, m_cfg.knnVal);
+  eOutput.reset();
+
+  ACTS_VERBOSE("Shape of built edges: (" << edgeList->size(0) << ", "
+                                         << edgeList->size(1));
+  ACTS_VERBOSE("Slice of edgelist:\n" << edgeList->slice(1, 0, 5));
+  // print_current_cuda_meminfo(logger);
+
+  // timeInfo.building = timer.stopAndGetElapsedTime();
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/TorchMetricLearning.cpp:94]: dead code?

File: Plugins/ExaTrkX/src/weaklyConnectedComponentsCugraph.hpp [hunk 16]
```diff
@@ -44,6 +44,7 @@ __global__ void weaklyConnectedComponents(std::vector<vertex_t>& rowIndices,
                                           std::vector<vertex_t>& colIndices,
                                           std::vector<weight_t>& edgeWeights,
                                           std::vector<vertex_t>& trackLabels) {
+#if 0
```
Comment (andiwand) [inline Plugins/ExaTrkX/src/weaklyConnectedComponentsCugraph.hpp:47]: intended shadowing?

**Review by andiwand — APPROVED** [review]
had a quick walk through - nothing critical from my side. as always feel free to just resolve my comment if it does not make sense

File: Examples/Python/python/acts/examples/reconstruction.py [hunk 17]
```diff
@@ -1193,28 +1193,51 @@ def addExaTrkX(
...
+
+        graphConstructor = acts.examples.TorchMetricLearning(**metricLearningConfig)
+        edgeClassifiers = [
+            acts.examples.TorchEdgeClassifier(**filterConfig),
+            acts.examples.TorchEdgeClassifier(**gnnConfig),
+        ]
+        trackBuilder = acts.examples.BoostTrackBuilding()
+    elif backend == ExaTrkXBackend.Onnx:
+        metricLearningConfig["modelPath"] = str(modelDir / "embedding.onnx")
+        filterConfig["modelPath"] = str(modelDir / "filtering.onnx")
```
[the remark this answers is not in the document]
Comment (benjaminhuth) [reply Examples/Python/python/acts/examples/reconstruction.py:1223]: yeah this depends on the files in the cloud storage... I don't want to change this in this PR

File: Examples/Algorithms/TrackFindingExaTrkX/src/TrackFindingAlgorithmExaTrkX.cpp [hunk 18]
```diff
@@ -24,14 +26,50 @@ ActsExamples::TrackFindingAlgorithmExaTrkX::TrackFindingAlgorithmExaTrkX(
...
+  if (!m_cfg.graphConstructor) {
+    throw std::invalid_argument("Missing graph construction module");
+  }
+  if (!m_cfg.trackBuilder) {
+    throw std::invalid_argument("Missing track building module");
+  }
+  if (m_cfg.edgeClassifiers.empty() or
+      not std::all_of(m_cfg.edgeClassifiers.begin(),
+                      m_cfg.edgeClassifiers.end(),
+                      [](const auto& a) { return static_cast<bool>(a); })) {
```
[the remark this answers is not in the document]
Comment (benjaminhuth) [reply Examples/Algorithms/TrackFindingExaTrkX/src/TrackFindingAlgorithmExaTrkX.cpp:36]: Yeah for some reason unknown to humanity the unary predicate is mandatory for `std::all_of`... Not sure why there is no overload for the case that the `element_type` is convertible to `bool`...

**Review by andiwand — APPROVED** [review]

**Pull request merged**